File: 029.Divide Two Integers.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        if (divisor == 0 || (divisor == -1 && dividend == INT_MIN))
            return INT_MAX;
        
        bool negative = false;
        
        long long longDividend;
        long long longDivisor;
        if (dividend < 0) {
            negative = !negative;
            longDividend = abs((long long)dividend);
        } else {
            longDividend = dividend;
        }
        
        if (divisor < 0) {
            negative = !negative;
            longDivisor = abs((long long) divisor);
        } else {
            longDivisor = divisor;
        }
            
        int result = 0;
        while (longDividend >= longDivisor) {
            long long multiple = 1;
            long long divTmp = longDivisor;
            while (longDividend > divTmp << 1) {
                divTmp = divTmp << 1;
                multiple = multiple << 1;
            }
            result += multiple;
            longDividend -= divTmp;
        }
        
        return negative? -result : result;
    }
};
==================== Time out version =========================
class Solution {
public:
    int divide(int dividend, int divisor) {
        bool negative = false;
        if (divisor == 0)
            return std::numeric_limits<int>::max();
        if (dividend < 0) {
            negative = !negative;
            if (dividend == -2147483648)
                dividend = 2147483647;
            else
                dividend = abs(dividend);
        }
        if (divisor < 0) {
            negative = !negative;
            divisor = abs(divisor);
        }
            
        int count = 0;
        while (dividend >= divisor) {
            count++;
            dividend -= divisor;
        }
        
        if (negative)
            count = 0 - count;
        
        return count;
    }
};

File: LeetCode/029.Divide Two Integers.py
class Solution(object):
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        MAX_INT= 2147483647
        if divisor == 0:
            return MAX_INT

        is_negative = False
        if dividend < 0:
            dividend = 0 - dividend
            is_negative = not is_negative
        if divisor < 0:
            divisor = 0 - divisor
            is_negative = not is_negative

        if divisor == 1 and not is_negative and dividend > MAX_INT:
            return MAX_INT
        
        divs = [divisor]
        cnts = [1]
        i = 0
        while divs[i] + divs[i] <= dividend:
            divs.append(divs[i] + divs[i])
            cnts.append(cnts[i] + cnts[i])
            i += 1

        count = 0
        i = len(cnts) - 1
        while dividend > 0 and i >= 0:
            if dividend - divs[i] >= 0:
                dividend -= divs[i]
                count += cnts[i]
            i -= 1

        if is_negative:
            return 0 - count
        return MAX_INT if count > MAX_INT else count
